hydroshare_jupyter_sync/lib/filesystem/fs_resource_map.py

```diff
@@ -0,0 +1,199 @@
+            self._insert(relative_resource_file)
+
+    def delete_file(self, relative_resource_file: Union[Path, str]) -> None:
+        if self._is_member(relative_resource_file):
+            relative_resource_file = self._as_child_of_base_directory(
+                relative_resource_file
+            )
+            del self.data[relative_resource_file]
+
+    def update_resource(self) -> None:
+        # clear data dictionary
+        self.data = dict()
+        for resource_file in self.data_path.glob("**/*"):
+            # insert relative file path to data_path and md5 digest
+            self._insert(resource_file)
+
+    @property
+    def base_directory(self) -> Path:
+        """Return assumed path to base directory (i.e. `/some/path/{resource_id}/{resource_id}`).
+
+        The property name, `base_directory`, was chosen based on nameing conventions found in any
+        HydroShare resource's `readme.txt`"""
+        return self.resource_path / self.resource_id
+
+    @property
+    def data_path(self):
+        """Return assumed path to resource data (i.e. `/some/path/{resource_id}/{resource_id}/data/contents`)"""
+        return self.base_directory / "data" / "contents"
+
+    # Helper methods
+    def _abs_path(self, resource_file: Union[Path, str]) -> Path:
+        resource_file = Path(resource_file)
+        return (
+            resource_file
+            if resource_file.is_absolute()
+            else self.base_directory / resource_file
+        ).resolve()
+
+    def _is_member(self, resource_file: Union[Path, str]) -> bool:
+        if self._is_child_of_data_path(resource_file):
+            return self._as_child_of_base_directory(resource_file) in self.data
+        return False
+
+    def _valid_resource_file(self, resource_file: Union[Path, str]) -> bool:
+        abs_path = self._abs_path(resource_file)
+
+        return (
+            abs_path.is_file()
+            and not abs_path.is_symlink()
+            and self._is_child_of_data_path(abs_path)
+        )
+
+    def _as_child_of(
+        self, resource_file: Union[Path, str], parents: Union[Path, str]
+    ) -> Path:
+        abs_path = self._abs_path(resource_file)
+        return abs_path.relative_to(parents)
+
+    def _as_child_of_data_path(self, resource_file: Union[Path, str]) -> Path:
+        return self._as_child_of(resource_file, self.data_path)
+
+    def _as_child_of_base_directory(self, resource_file: Union[Path, str]) -> Path:
+        return self._as_child_of(resource_file, self.base_directory)
+
+    def _is_child_of_data_path(self, resource_file: Union[Path, str]) -> bool:
+        try:
+            # ensure is child of data_path
+            self._as_child_of_data_path(resource_file)
+            return True
+        except ValueError:
+            return False
+
+    def _insert(self, resource_file: Union[Path, str]) -> None:
+        abs_path = self._abs_path(resource_file)
+
+        if self._valid_resource_file(abs_path):
```

Comment:
perhaps instead of raising a custom error, you can log a WARN when your condition is not met.  Or maybe an assert instead of if?